rfcs/20200624-pluggable-device-for-tensorflow.md

```diff
@@ -0,0 +1,349 @@
+<div align="center">
+<img src=20200624-pluggable-device-for-tensorflow/scenario1.png>
+</div>  
+
+* **Supported scenario**: Single PluggableDevice registered as a new device type.  
+  In the case of installing one plugin that registers its PluggableDevice as a new device type, e.g., "XPU" device type, user can speficify the "XPU" device for ops under `with tf.device("xpu:0")`, PluggableDevice registered will be selected to run those ops.
+<div align="center">
+<img src=20200624-pluggable-device-for-tensorflow/scenario2.png>
+</div>
+
+* **Supported scenario**: Multiple PluggableDevices registered as different device types.   
+  In the case of installing multiple plugins that register PluggableDevices as different device types, e.g., one is registered as "GPU" device type and another is registered as "XPU" device type, these PluggableDevices can be registered successfully and user can specify the registered device type to run ops on different hardware. Same as scenario1, when one PluggableDevice is registered as "GPU" device type, the default GPUDevice will be overrided.
+<div align="center">
+<img src=20200624-pluggable-device-for-tensorflow/scenario3.png>
+</div>
+
+* **Non-Supported scenario**: Multiple PluggableDevices registered as the same device type.  
+  In the case of installing multiple plugins that registers PluggableDevice as the same device type, e.g., more than one plugin registers its PluggableDevice as "GPU" device type, these plugins's initialization will fail due to registration conflict. User needs to manually select which platform they want to use(either unloads the conflicting plugin or reconfigures the plugin with python API).
+<div align="center">
+<img src=20200624-pluggable-device-for-tensorflow/scenario4.png>
+</div>
+
+### Front-end Mirroring mechanism
+This section describes the front-end mirroring mechanism for python users, pointing at previous user scenarios.
+* **device type**  
+   Device type is user visible and controllable. User can specify the device type for the ops. e.g, specify device type as "gpu", "xpu".
+   ```
+   >> with tf.device("/gpu:0"):
+      ...
+   >> with tf.device("/xpu:0"):
+      ...
+   ```
+   user need to manually select a platform when multiple plugins register the same device type, or the plugins initilaization will fail due to device type conflict. e.g, when multiple plugins register the "GPU" device type ,user need to set a higher priority for the plugin.
+   ```
+   >> tf.load_plugin_with_highest_priority(path_to_plugin_lib)
+   >> with tf.device("/gpu:0")
+      ...
+   ```
+* **subdevice type**  
+   Subdevice type is user visible but not controllable. User can query the subdevice type of the device if he wants to know whether the GPU device is NVIDIA_GPU or INTEL_GPU, through [tf.config.experimental.list_physical_devices()](https://www.tensorflow.org/api_docs/python/tf/config/list_physical_devices).
```

Comment:
hi, @alextp in the case of installing a plugin that registers the "GPU" device type, which front-end usage is preferred?
**option1**:  
when user list_physical_devices(), only higher priority device can be visible, default GPUDevice will be invalid. when user what to use default GPUDevice, he need to uninstall the plugin.
```
>> gpu_device = tf.config.experimental.list_physical_devices(`GPU`)
   >> print(gpu_device)
   [PhysicalDevice(name = `physical_device:GPU:0`), device_type = `GPU`, subdevice_type = `INTEL_GPU`]
```
**option2**:    
when user list_physical_devices(), all the gpu device is listed, PluggableDevice is default used. when user wants to use the default GPUDevice, a new API will be introduced to mirroring. However, in this usage, there is an 
drawback: may have semantic change with exsiting api, such as  example of [list_physical_devices](https://www.tensorflow.org/api_docs/python/tf/config/list_physical_devices) , the num of gpus is not correct
```
>> gpu_device = tf.config.experimental.list_physical_devices(`GPU`)
   >> print(gpu_device)
   [PhysicalDevice(name = `physical_device:GPU:0`), device_type = `GPU`, subdevice_type = `INTEL_GPU`, used]
   [PhysicalDevice(name = `physical_device:GPU:0`), device_type = `GPU`, subdevice_type = `NVIDIA_GPU`, unused]
```  
example in [tf.config.list_physical_devices](https://www.tensorflow.org/api_docs/python/tf/config/list_physical_devices)  
```
physical_devices = tf.config.list_physical_devices('GPU')
print("Num GPUs:", len(physical_devices))
```